blobstore/src/main/java/org/jclouds/blobstore/util/BlobStoreUtils.java

```diff
@@ -85,4 +103,65 @@ private static void convertUserMetadataKeysToLowercase(MutableBlobMetadata metad
       }
       metadata.setUserMetadata(lowerCaseUserMetadata);
    }
+
+   private static class CrawlBlobStoreIterable implements Iterable<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+
+      CrawlBlobStoreIterable(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options);
+      }
+
+      @Override
+      public Iterator<StorageMetadata> iterator() {
+         return new CrawlBlobStoreIterator(blobStore, containerName, options);
+      }
+   }
+
+   private static class CrawlBlobStoreIterator implements Iterator<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+      private Iterator<? extends StorageMetadata> iterator;
+      private String marker;
+
+      CrawlBlobStoreIterator(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options).recursive();
+         advance();
+      }
+
+      private void advance() {
+         if (marker != null) {
+            options.afterMarker(marker);
+         }
+         PageSet<? extends StorageMetadata> set = blobStore.list(containerName, options);
+         marker = set.getNextMarker();
+         iterator = set.iterator();
+      }
+
+      @Override
+      public boolean hasNext() {
+         return iterator.hasNext() || marker != null;
+      }
+
+      @Override
+      public StorageMetadata next() {
+         while (true) {
+            if (!iterator.hasNext()) {
+               advance();
+            }
+            StorageMetadata metadata = iterator.next();
```

Comment:
Changes lgtm now. However, advancing the iterator in the `hasNext` is not the best pattern to use. I'd suggest to use the Guava's `AbstractIterator` to implement the iterator. You can take a look at the GCE's [AdvancingIterator](https://github.com/jclouds/jclouds-labs-google/blob/master/googlecloud/src/main/java/org/jclouds/googlecloud/internal/AdvancingIterator.java), which does pretty much the same. Your `crawl` methods should use that iterator and the [this concat method](https://github.com/jclouds/jclouds-labs-google/blob/master/googlecloud/src/main/java/org/jclouds/googlecloud/internal/ListPages.java#L34-44) to return an Iterator that can advance between pages, without having to change the state of the iterator in the `hasNext` method. WDYT?